cli/azd/pkg/auth/manager.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err := m.saveLoginForPublicClient(res); err != nil {
+		return nil, err
+	}
+
+	return newAzdCredential(m.publicClient, &res.Account), nil
+}
+
+func (m *Manager) LoginWithDeviceCode(ctx context.Context, deviceCodeWriter io.Writer) (azcore.TokenCredential, error) {
+	code, err := m.publicClient.AcquireTokenByDeviceCode(ctx, cLoginScopes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Display the message to the end user as to what to do next, then block waiting for them to complete
+	// the flow.
+	fmt.Fprintln(deviceCodeWriter, code.Result.Message)
+
+	res, err := code.AuthenticationResult(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := m.saveLoginForPublicClient(res); err != nil {
+		return nil, err
+	}
+
+	return newAzdCredential(m.publicClient, &res.Account), nil
+
+}
+
+func (m *Manager) LoginWithServicePrincipalSecret(
+	ctx context.Context, tenantId, clientId, clientSecret string,
+) (azcore.TokenCredential, error) {
+	cred, err := azidentity.NewClientSecretCredential(tenantId, clientId, clientSecret, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating credential: %w", err)
+	}
+
+	if err := m.saveLoginForServicePrincipal(
+		tenantId,
+		clientId,
+		&persistedSecret{
+			ClientSecret: &clientSecret,
+		},
+	); err != nil {
+		return nil, err
+	}
+
+	return cred, nil
+}
+
+func (m *Manager) LoginWithServicePrincipalCertificate(
+	ctx context.Context, tenantId, clientId string, certData []byte,
+) (azcore.TokenCredential, error) {
+	certs, key, err := azidentity.ParseCertificates(certData, nil)
+	if err != nil {
+		return nil, fmt.Errorf("parsing certificate: %w", err)
+	}
+
+	cred, err := azidentity.NewClientCertificateCredential(tenantId, clientId, certs, key, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating credential: %w", err)
+	}
+
+	encodedCert := base64.StdEncoding.EncodeToString(certData)
+
+	if err := m.saveLoginForServicePrincipal(
+		tenantId,
+		clientId,
+		&persistedSecret{
+			ClientCertificate: &encodedCert,
+		},
+	); err != nil {
+		return nil, err
+	}
+
+	return cred, nil
+}
+
+// Logout signs out the current user and removes any cached authentication information
+func (m *Manager) Logout(ctx context.Context) error {
+	act, err := m.getSignedInAccount(ctx)
+	if err != nil && !errors.Is(err, ErrNoCurrentUser) {
+		return fmt.Errorf("fetching current user: %w", err)
+	}
+
+	if act != nil {
+		if err := m.publicClient.RemoveAccount(*act); err != nil {
+			return fmt.Errorf("removing account from msal cache: %w", err)
+		}
+	}
+
+	cfg, err := config.GetUserConfig(m.configManager)
+	if err != nil {
+		return fmt.Errorf("loading config: %w", err)
+	}
+
+	// we are fine to ignore the error here, it just means there's nothing to clean up.
+	currentUser, _ := readUserProperties(cfg)
+
+	if currentUser != nil {
+		// When logged in as a service principal, remove the stored credential
+		if currentUser.TenantID != nil && currentUser.ClientID != nil {
```

Comment:
nit: Consider combining these if expressions instead of having them nested.

Reply:
Yes, that's much better.  Thanks for catching that.